#' Simple saving function
#'
#' @param x Df to save
#' @param y save name
#' @param z Group name.
#' @param analysis_folder main folder hosting analsysis
#' @return Saves data frames
#' @importFrom magrittr %>%
#' @importFrom rlang .data
saving_files_simple <- function(x,y,z,analysis_folder) {
  file_to_save <- x
  data.table::fwrite(file_to_save,paste0(analysis_folder,"/","split_",z,"/",y),sep = "\t", na = NA)
}
#' edits plink results
#'
#' @param x result files
#' @param y name of files
#' @param remove_string vector to remove
#' @param analysis_name name of the analysis
#' @return Saves data frames
#' @importFrom magrittr %>%
#' @importFrom rlang .data
plink_file_edit_join <- function(x,y,remove_string,analysis_name) {
  name <- data.frame(stringr::str_remove_all(y, paste(remove_string, collapse = "|"))) %>%
    dplyr::rename(PheWAS_ID=1)
  per_result <- data.table::fread(x) %>%
    dplyr::bind_cols(name) %>%
    dplyr::mutate(dplyr::across(tidyselect::everything(), as.character),
                  table_save=analysis_name)
  return(per_result)
}

#' Runs Plink2 association testing for the extracted variants
#'
#' @param x Phenotype data frame
#' @param y Group name
#' @param split_groups Group names(s) that are to be split
#' @param analysis_folder folder hosting the outputs of the function
#' @param psam the psam file of the extracted SNPs
#' @param N_binary_split numerical value for how many binary phenotypes to include in each split.
#' @param N_quant_split numerical value for how many quantitative phenotypes to include in each split.
#' @param binary_ID Binary phenotype PheWAS_IDs
#' @param quant_ID Quantitative phenotype PheWAS_IDs
#' @param check_existing_results logical value used whether to look for existing Plink results in analysis folder structure and to not repeat those associations.
#' @param save_plink_tables all available phenotypes generated by the previous functions.
#' @param analysis_name name of the analysis for saving.
#' @param plink_exe required input in system to run Plink2
#' @param association_variants extracted SNPs
#' @param model genetic model used in Plink
#' @param covariates covariates file for association analysis
#' @param split_analysis logical value if split analysis is being used.
#' @return Plink association results.
#' @importFrom magrittr %>%
#' @importFrom rlang .data
association_tests <- function(y,x,
                              split_groups,
                              analysis_folder,
                              psam,
                              N_binary_split,
                              N_quant_split,
                              binary_ID,
                              quant_ID,
                              check_existing_results,
                              save_plink_tables,
                              analysis_name,
                              plink_exe,
                              association_variants,
                              model,
                              covariates,
                              split_analysis) {
  . <- NULL
  if(any(stringr::str_detect(split_groups,y))) {
    #create directory for the split files
    dir.create(paste0(analysis_folder,"/","split_",y))
    # edit phenotype file id much like covariates
    phenotype_edit <- data.table::fread(x) %>%
      dplyr::rename(IID=1) %>%
      dplyr::filter(.data$IID %in% psam$IID)
    # define split values for different types of phenotypes
    if(is.na(as.numeric(N_binary_split))){
      rlang::abort(paste0("'N_binary_split' must be a numeral"))
    }
    split_number_binary <- as.numeric(N_binary_split)
    if(is.na(as.numeric(N_quant_split))){
      rlang::abort(paste0("'N_quant_split' must be a numeral"))
    }
    split_number_quant <- as.numeric(N_quant_split)
    # separate phenotypes into quant and binary
    binary_phenotypes <- phenotype_edit %>%
      dplyr::select(1,tidyselect::any_of(binary_ID))
    quant_phenotypes <- phenotype_edit %>%
      dplyr::select(1,tidyselect::any_of(quant_ID))
    # now split PheWAS_ID by the values selected
    binary_column_splits <- split(colnames(binary_phenotypes)[-1], ceiling(seq_along(colnames(binary_phenotypes)[-1])/split_number_binary))
    quant_column_splits <- split(colnames(quant_phenotypes)[-1], ceiling(seq_along(colnames(quant_phenotypes)[-1])/split_number_quant))
    # combine
    column_splits <- c(binary_column_splits,quant_column_splits)
    # use PheWAS_ID splits to select off columns
    split_phenotypes_df <- lapply(column_splits, function(x) split_pheno <- phenotype_edit %>%
                                    dplyr::select(1,tidyselect::any_of(x)))
    # create a file for each of the splits
    N_splits <- length(split_phenotypes_df)
    file_names <- paste0(y,"_",seq(1:N_splits))
    #save
    mapply(saving_files_simple,
           split_phenotypes_df,
           file_names,
           y,
           MoreArgs = list(analysis_folder=analysis_folder))
    # guide to help with distributed computing
    guide_file <- data.frame(list.files(paste0(analysis_folder,"/","split_",y),full.names = T))
    data.table::fwrite(guide_file,paste0(analysis_folder,"/",y,"_split_guide"),col.names = F)

    return()
  }
  #save location for the plink results
  save_location <- paste0(analysis_folder,"/","plink_results","/",y)
  group_name <- y
  #check existing files for phenotypes
  all_pheno_results <- list.files(paste0(analysis_folder,"/","plink_results"))
  convert_PheWAS_ID <- stringr::str_remove(stringr::str_remove(all_pheno_results[stringr::str_detect(all_pheno_results,y)],paste0(y,".")),".glm.*")
  # ability to just run pheno-group test that havn't already been completed
  if(check_existing_results){
    remaining_binary <- setdiff(binary_ID,convert_PheWAS_ID)
    remaining_quatitative <- setdiff(quant_ID,convert_PheWAS_ID)
    all_remaining <- c(remaining_binary,remaining_quatitative)
  } else {
    all_remaining <- c(binary_ID,quant_ID)
  }
  # edit phenotype file id much like covariates
  phenotype_edit <- data.table::fread(x) %>%
    dplyr::rename(IID=1) %>%
    dplyr::filter(.data$IID %in% psam$IID) %>%
    dplyr::select(.data$IID,tidyselect::any_of(all_remaining))
  # exit if no phenotypes to test
  if(ncol(phenotype_edit)==1){
    if(isFALSE(split_analysis)){
      # combine results
      results_files <-  list.files(path = paste0(analysis_folder,"/","plink_results"), pattern = paste0(y,"(.*?)hybrid|",y,"(.*?)linear"))
      remove_string <- c(".glm.logistic.hybrid", paste0(y,"."), ".glm.linear")
      results_file_location <- paste0(paste0(analysis_folder,"/","plink_results","/",results_files))

      # join the plink results into a single table
      plink_results <- mapply(plink_file_edit_join,
                              results_file_location,
                              results_files,
                              MoreArgs = list(remove_string=remove_string,analysis_name=analysis_name),
                              SIMPLIFY = F) %>%
        data.table::rbindlist(.,fill = T)
      # option to save raw results per group
      if(save_plink_tables){
        save_results_table_name <- paste0(y,"_",analysis_name,"_plink_results_raw.csv")
        data.table::fwrite(plink_results,paste0(analysis_folder,"/","association_results","/",save_results_table_name), na = NA)
      }
      return(plink_results)
    } else {
      return()
    }
  }
  # separate into binary and quant this is for neater results
  binary_pheno <- phenotype_edit %>%
    dplyr::mutate(`#FID`=.data$IID) %>%
    dplyr::select(.data$`#FID`,.data$IID,tidyselect::any_of(binary_ID))
  # quant
  quant_pheno <- phenotype_edit %>%
    dplyr::mutate(`#FID`=.data$IID) %>%
    dplyr::select(.data$`#FID`,.data$IID,tidyselect::any_of(quant_ID))
  # save locations for temp phenotypes
  binary_pheno_file <- paste0(analysis_folder,"/","temp_pheno","/",basename(x),"_temp_binary")
  quant_pheno_file <- paste0(analysis_folder,"/","temp_pheno","/",basename(x),"_temp_quant")
  # selecting whether to do binary or quant or both types of analysis
  if(ncol(quant_pheno)==2) {
    # write temp binary file
    data.table::fwrite(binary_pheno,binary_pheno_file, sep = "\t", na = NA, quote = F)
    # run the regression in plink2
    if(is.null(covariates)){
      system(paste0(plink_exe," --pfile ",association_variants," --glm allow-no-covars sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",binary_pheno_file," --ci 0.95  --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    } else {
      system(paste0(plink_exe," --pfile ",association_variants," --glm sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",binary_pheno_file," --covar ",covariates," --ci 0.95  --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    }
    # remove the binary file
    file.remove(binary_pheno_file)
  } else if(ncol(binary_pheno)==2) {
    data.table::fwrite(quant_pheno,quant_pheno_file, sep="\t", na = NA,quote = F)
    # run the regression in plink2
    if(is.null(covariates)){
      system(paste0(plink_exe," --pfile ",association_variants," --glm allow-no-covars sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",quant_pheno_file," --ci 0.95 --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    } else {
      system(paste0(plink_exe," --pfile ",association_variants," --glm sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",quant_pheno_file," --covar ",covariates," --ci 0.95 --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    }
    # remove the quant file
    file.remove(quant_pheno_file)
  } else {
    # write temp phenotype files
    data.table::fwrite(binary_pheno,binary_pheno_file, sep = "\t", na = NA, quote = F)
    data.table::fwrite(quant_pheno,quant_pheno_file, sep="\t", na = NA, quote = F)
    # run the regression in plink2
    if(is.null(covariates)){
      system(paste0(plink_exe," --pfile ",association_variants," --glm allow-no-covars sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",quant_pheno_file," --ci 0.95 --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
      system(paste0(plink_exe," --pfile ",association_variants," --glm allow-no-covars sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",binary_pheno_file," --ci 0.95  --1 --vif 90 --out ",save_location," --covar-variance-standardize"))

    } else {
    system(paste0(plink_exe," --pfile ",association_variants," --glm sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",binary_pheno_file," --covar ",covariates," --ci 0.95 --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    system(paste0(plink_exe," --pfile ",association_variants," --glm sex cols=+totallele,+a1count,+a1countcc,+a1countcc,+a1freq,+machr2,+a1freqcc hide-covar ",model," --pheno ",quant_pheno_file," --covar ",covariates," --ci 0.95 --1 --vif 90 --out ",save_location," --covar-variance-standardize"))
    }
    # remove the temp phenotype files
    file.remove(binary_pheno_file)
    file.remove(quant_pheno_file)
  }
  if(isFALSE(split_analysis)){
    # combine results
    results_files <-  list.files(path = paste0(analysis_folder,"/","plink_results"), pattern = paste0(y,"(.*?)hybrid|",y,"(.*?)linear"))
    remove_string <- c(".glm.logistic.hybrid", paste0(y,"."), ".glm.linear")
    results_file_location <- paste0(paste0(analysis_folder,"/","plink_results","/",results_files))

    # join the plink results into a single table
    plink_results <- mapply(plink_file_edit_join,
                            results_file_location,
                            results_files,
                            MoreArgs = list(remove_string=remove_string, analysis_name=analysis_name),
                            SIMPLIFY = F) %>%
      data.table::rbindlist(.,fill = T)
    # option to save raw results per group
    if(save_plink_tables){
      save_results_table_name <- paste0(y,"_",analysis_name,"_plink_results_raw.csv")
      data.table::fwrite(plink_results,paste0(analysis_folder,"/","association_results","/",save_results_table_name), na = NA)
    }
    return(plink_results)
  } else {
    return()
  }
}

#' Performs association testing using PLINK2.
#'
#' @param analysis_folder Full path of the directory in which the association results will be saved.
#' @param phenotype_files Comma-separated list containing the full paths of the phenotype data.
#' @param covariate Full path of a file containing covariate data to be used in model adjustment. See user guide for format.
#' @param variants_for_association Full path of the genetic data for the SNPs of interest (produced by the extract_SNP.R script).
#' @param analysis_name Name for the analysis to be used in the naming of the output files.
#' @param group_name_overide Comma-separated list containing alternative group names. By default, group names are extracted from the suffix of the file names provided in the phenotype_files argument. For example, if a file named /home/phenotypes/EUR_phenotypes.csv were provided, the corresponding group name would be "EUR". This argument allows for a different group name to be specified. The order of names provided should match the files specified in the phenotype_files argument.
#' @param PheWAS_manifest_overide Full file path of the alternative PheWAS_manifest file.
#' @param plink_exe Command to execute plink2 program.
#' @param save_plink_tables Select if wanting to save the raw plink results per group as a table always saved in analysis_folder/assoiation_results/group/group_plink_results_raw.
#' @param phenotype_inclusion_file Full file path to a plain txt file containing single column NO header containing full PheWAS_ID of phenotypes that will be included. Cannot be used with phenotype_exclusion_file argument.
#' @param phenotype_exclusion_file Full file path to a plain txt file containing single column NO header containing full PheWAS_ID of phenotypes that will be excluded. Cannot be used with phenotype_inclusion_file argument.
#' @param split_group Comma separated groups that require splitting for more efficient analysis. Does not run the analysis per group but splits and saves the phenotypes into smaller files which can then be analysed using cluster based computing. The number of phenotypes in each split file is dependent on the type (binary or quantitative) and is set using N_quant_split and N_binary_split. Split files are saved in analysis_folder/group_split with group being the group name. A file is created in analysis_folder/group_split saved as group_split_guide with group once again being the group name inputted. This file lists the file names of the splits, which can be used to guide distributed computing analysis.
#' @param N_quant_split Number of quantitative phenotypes per split file.
#' @param N_binary_split Number of binary phenotypes per split file.
#' @param split_analysis Select if the input being analysed is from the split_group command, only changes the saving and attempted combining of files.
#' @param model Genetic model to use for analysis, can be one of genotypic, hethom, dominant, recessive, hetonly. If not one of these options will be converted to blank.
#' @param check_existing_results Input if wanting to re-run part of the analysis whilst checking for existing results in the plink_results folder.Used primarily where a run has aborted part way through.
#' @return Association test results.
#' @importFrom magrittr %>%
#' @importFrom rlang .data
#' @export
plink_association_testing <- function(analysis_folder,
                                      phenotype_files,
                                      covariate,
                                      variants_for_association,
                                      analysis_name,
                                      group_name_overide,
                                      PheWAS_manifest_overide,
                                      plink_exe,
                                      save_plink_tables,
                                      phenotype_inclusion_file,
                                      phenotype_exclusion_file,
                                      split_group,
                                      N_quant_split,
                                      N_binary_split,
                                      split_analysis,
                                      model,
                                      check_existing_results){
  # load in
  if(is.null(PheWAS_manifest_overide)){
    PheWAS_manifest <- data.table::fread(system.file("extdata","PheWAS_manifest.csv.gz", package = "DeepPheWAS"))
  } else {
    if(!file.exists(PheWAS_manifest_overide)){
      rlang::abort(paste0("'PheWAS_manifest_overide' must be a file"))
    }
    PheWAS_manifest <- data.table::fread(PheWAS_manifest_overide)
    PheWAS_manifest_overide_colname <- c("PheWAS_ID","broad_catagory","category","phenotype","range_ID","sex","field_code","QC_flag_ID","QC_filter_values","first_last_max_min_value","date_code","age_code","case_code","control_code","exclude_case_control_crossover","included_in_analysis","quant_combination","primary_care_code_list","limits","lower_limit","upper_limit","transformation","analysis","age_col","pheno_group","short_desc","group_narrow","concept_name","concept_description","Notes")

    if(!all(tibble::has_name(PheWAS_manifest,PheWAS_manifest_overide_colname))){
      warning(paste0("'PheWAS_manifest_overide' does not have the correct colnames and may not produce the correct output, expected colnames are:
                                 '"),paste(PheWAS_manifest_overide_colname, collapse=","),paste0("'
                                 not:
                                 "),paste(colnames(PheWAS_manifest), collapse=","),
              paste0("
                                 differences between inputed file and expected are:
                                 "),paste(setdiff_all(names(PheWAS_manifest),PheWAS_manifest_overide_colname), collapse=","))
    }
  }
  # creating analysis folder
  if(!dir.exists(paste0(analysis_folder,"/","association_results"))) {
    dir.create(paste0(analysis_folder,"/","association_results"),recursive = T)
  }
  #create folder for temp phenotypes
  if(!dir.exists(paste0(analysis_folder,"/","temp_pheno"))){
    dir.create(paste0(analysis_folder,"/","temp_pheno"),recursive = T)
  }
  # assigning groups and phenotype_files dependent on input
  phenotype_files <- unlist(stringr::str_split(phenotype_files, ","))
  if(!is.null(group_name_overide)) {
    groups <- unlist(stringr::str_split(group_name_overide, ","))
  } else {
    groups <-gsub("_.*", "", basename(unlist(strsplit(phenotype_files,","))))
  }

  if(!is.null(split_group)) {
    split_groups <- unlist(stringr::str_split(split_group,pattern = ","))
  } else {
    split_groups <- ""
  }

  # variants for association edit to allow input into plink2
  association_variants <- stringr::str_remove(variants_for_association,".pgen")
  # defining all_phenos
  additional_phenos <- PheWAS_manifest %>%
    dplyr::filter(.data$included_in_analysis==1) %>%
    dplyr::mutate(male=paste0(.data$PheWAS_ID,"_male"),
                  female=paste0(.data$PheWAS_ID,"_female"),
                  age_of_onset=paste0(.data$PheWAS_ID,"_age_of_onset"),
                  male_analysis=.data$analysis,
                  female_analysis=.data$analysis,
                  age_of_onset_analysis="quant",
                  male_age_col=.data$age_col,
                  female_age_col=.data$age_col,
                  age_of_onset_age_col="age") %>%
    dplyr::select(.data$PheWAS_ID,.data$male,.data$female,.data$analysis,.data$male_analysis,.data$female_analysis,.data$age_col,.data$male_age_col,.data$female_age_col,.data$age_of_onset,.data$age_of_onset_analysis,.data$age_of_onset_age_col) %>%
    dplyr::mutate(age_col_complete = ifelse(.data$age_col=="named",paste0(.data$PheWAS_ID,"_age"),"age"))

  all_possible_phenotypes <- data.frame(PheWAS_ID=c(additional_phenos$PheWAS_ID,additional_phenos$male,additional_phenos$female,additional_phenos$age_of_onset),
                                        analysis=c(additional_phenos$analysis,additional_phenos$male_analysis,additional_phenos$female_analysis,additional_phenos$age_of_onset_analysis),
                                        age_col=c(additional_phenos$age_col,additional_phenos$male_age_col,additional_phenos$female_age_col,additional_phenos$age_of_onset_age_col))

  if(!is.null(phenotype_inclusion_file)){
    if(!file.exists(phenotype_inclusion_file)){
      rlang::abort(paste0("'phenotype_inclusion_file' must be a file"))
    }
    phenotype_inclusion_ID <- data.table::fread(phenotype_inclusion_file, header = F) %>%
      dplyr::pull(1)
    all_phenos <- all_possible_phenotypes %>%
      dplyr::filter(.data$PheWAS_ID %in% phenotype_inclusion_ID)
  } else if (!is.null(phenotype_exclusion_file)){
    if(!file.exists(phenotype_exclusion_file)){
      rlang::abort(paste0("'phenotype_exclusion_file' must be a file"))
    }
    phenotype_exclusion_ID <- data.table::fread(phenotype_exclusion_file, header = F) %>%
      dplyr::pull(1)
    all_phenos <- all_possible_phenotypes %>%
      dplyr::filter(!.data$PheWAS_ID %in% phenotype_exclusion_ID)
  } else {
    all_phenos <- all_possible_phenotypes
  }

  # create required folder
  if(!dir.exists(paste0(analysis_folder,"/","plink_results"))) {
    dir.create(paste0(analysis_folder,"/","plink_results"))
  }
  # split phenotypes
  binary_ID <- all_phenos %>%
    dplyr::filter(.data$analysis=="binary") %>%
    dplyr::pull(.data$PheWAS_ID)
  quant_ID <- all_phenos %>%
    dplyr::filter(.data$analysis=="quant") %>%
    dplyr::pull(.data$PheWAS_ID)
  # psam required to match sample id for testing
  if(!file.exists(paste0(association_variants,".psam"))){
    rlang::abort(paste0("'psam' file required"))
  }
  psam <- data.table::fread(paste0(association_variants,".psam"))
  # edited to match psam and required col names
  if(!is.null(covariate)){
    if(!file.exists(covariate)){
      rlang::abort(paste0("'covariate' file required"))
    }
    edited_covartiate <- data.table::fread(covariate) %>%
      dplyr::filter(.data$IID %in% psam$IID)
    data.table::fwrite(edited_covartiate,paste0(analysis_folder,"/","edited_covariate"),na = NA,quote = F,sep = "\t")
    covariates <- paste0(analysis_folder,"/","edited_covariate")
  } else {
    covariates <- NULL
  }
  # location for plink2 commands
  plink_exe <- plink_exe
  if(!is.null(model)){
    model <- model
    acceptable_input <- c("recessive","dominant","hetonly","genotypic","hethom")
    if(!model %in% acceptable_input){
      warning(paste0("'model' is not one of the accepted inputs:
                                 '"),paste(acceptable_input,collapse=" , "))
      model <- ""
    }
  } else {
    model <- ""
  }
  # run associations
  all_results <- mapply(association_tests,
                        groups,
                        phenotype_files,
                        MoreArgs = list(split_groups=split_groups,
                                        analysis_folder=analysis_folder,
                                        psam=psam,
                                        N_binary_split=N_binary_split,
                                        N_quant_split=N_quant_split,
                                        binary_ID=binary_ID,
                                        quant_ID=quant_ID,
                                        check_existing_results=check_existing_results,
                                        save_plink_tables=save_plink_tables,
                                        analysis_name=analysis_name,
                                        plink_exe=plink_exe,
                                        association_variants=association_variants,
                                        model=model,
                                        covariates=covariates,
                                        split_analysis=split_analysis),
                        SIMPLIFY = F, USE.NAMES = T)
  #save results
  if(split_analysis) {
  } else {
    saveRDS(all_results,paste0(analysis_folder,"/","association_results","/",analysis_name,"_association_results_list.gz"))
  }
}
